### PR TITLE
Fix rehash functionality

### PR DESCRIPTION
**Pull request by chrisseaton**
Sorry, @mattalp is busy at the moment, and his work was on a private branch. This replaces #2266. Sorry, please continue the discussion here. I've fixed everything except https://github.com/oracle/truffleruby/pull/2266#issuecomment-812952534.

### CODE REVIEW

**Comment by chrisseaton** [comment]
Fixes #2263.

File: src/main/java/org/truffleruby/core/hash/HashNodes.java [hunk 1]
```diff
@@ -886,28 +897,44 @@ protected RubyHash rehashBuckets(RubyHash hash,
...
+                    Entry previousEntry = entry;
+
+                    do {
+                        if (compareHashKeysNode.equalKeys(
+                                compareByIdentity,
+                                entry.getKey(),
+                                newHash,
+                                bucketEntry.getKey(),
+                                bucketEntry.getHashed())) {
+                            BucketsStrategy.removeFromSequenceChain(hash, entry);
```
Comment (eregon) [inline src/main/java/org/truffleruby/core/hash/HashNodes.java:925]: It seems weird and error-prone that `removeFromSequenceChain()` only does half of it.
I think it should do both, and `DeleteLastNode` should instead use a variant of it (`removeLastFromSequenceChain`, which asserts `entry.getNextInSequence() == null`).

**Review by eregon — COMMENTED** [review]
Looks good.
We need a spec to make sure the behavior is correct though.

Could you add something like `it "removes duplicate keys for large hashes"` and then test a hash with 10 entries?

**Comment by chrisseaton** [comment]
Updated.

File: src/main/java/org/truffleruby/core/hash/BucketsStrategy.java [hunk 2]
```diff
@@ -194,14 +194,17 @@ public static void copyInto(RubyContext context, RubyHash from, RubyHash to) {
...
         if (entry.getPreviousInSequence() == null) {
             assert hash.firstInSequence == entry;
             hash.firstInSequence = entry.getNextInSequence();
         } else {
             assert hash.firstInSequence != entry;
             entry.getPreviousInSequence().setNextInSequence(entry.getNextInSequence());
+            final Entry nextInSequence = entry.getNextInSequence();
+            if (nextInSequence != null) {
+                nextInSequence.setPreviousInSequence(entry.getPreviousInSequence());
+            }
```
Comment (eregon) [inline src/main/java/org/truffleruby/core/hash/BucketsStrategy.java:207]: There is missing logic here to update `hash.lastInSequence`, I'll fix it. With that then it's perfectly symmetrical.

**Review by eregon — APPROVED** [review]
Thank you for the fix!

**Pull request merged**